ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LedgerSeq.hs

```diff
@@ -0,0 +1,479 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE QuantifiedConstraints      #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+-- | The data structure that holds the cached ledger states.
+module Ouroboros.Consensus.Storage.LedgerDB.V2.LedgerSeq (
+    -- * LedgerHandles
+    LedgerTablesHandle (..)
+    -- * The ledger seq
+  , LedgerSeq (..)
+  , LedgerSeq'
+  , StateRef (..)
+  , closeLedgerSeq
+  , empty
+  , empty'
+    -- * Apply Blocks
+  , applyBlock
+  , applyThenPush
+  , extend
+  , prune
+  , pruneToImmTipOnly
+    -- * Queries
+  , anchor
+  , current
+  , currentHandle
+  , getPastLedgerAt
+  , immutableTipSlot
+  , isSaturated
+  , maxRollback
+  , rollback
+  , rollbackN
+  , rollbackToAnchor
+  , rollbackToPoint
+  , snapshots
+  , tip
+  , volatileStatesBimap
+  ) where
+
+import           Data.Function (on)
+import           Data.Word
+import           GHC.Generics
+import           NoThunks.Class
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config.SecurityParam
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.Extended
+import           Ouroboros.Consensus.Ledger.Tables.Utils
+import           Ouroboros.Consensus.Storage.LedgerDB.Config
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Network.AnchoredSeq hiding (anchor, last, map,
+                     rollback)
+import qualified Ouroboros.Network.AnchoredSeq as AS hiding (map)
+import           Prelude hiding (read)
+import           System.FS.API
+import           System.FS.API.Types
+
+{-------------------------------------------------------------------------------
+  LedgerTablesHandles
+-------------------------------------------------------------------------------}
+
+data LedgerTablesHandle m l = IOLike m => LedgerTablesHandle {
+    close       :: !(m ())
+  , duplicate   :: !(m (LedgerTablesHandle m l))
+  , read        :: !(LedgerTables l KeysMK -> m (LedgerTables l ValuesMK))
+  , readAll     :: !(m (LedgerTables l ValuesMK))
```

Comment:
Range lookups for LSM-tree style handles would require a range of keys as an argument (see [the docs](https://input-output-hk.github.io/lsm-tree/lsm-tree/Database-LSMTree-Normal.html#t:Range)). Reading all keys at once would be possible, but that means we would be reading the full UTxO into memory which would greatly increase memory use, so probably we should use a range of keys here too

Reply:
Defered to next PR